frontend/src/main/java/com/google/tripmeout/frontend/servlets/PlaceParentServlet.java

```diff
@@ -0,0 +1,106 @@
+package com.google.tripmeout.frontend.servlets;
+
+import com.google.common.flogger.FluentLogger;
+import com.google.gson.Gson;
+import com.google.maps.GeoApiContext;
+import com.google.maps.PlaceDetailsRequest;
+import com.google.maps.errors.InvalidRequestException;
+import com.google.maps.model.PlaceDetails;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PlaceParentServlet extends HttpServlet {
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+  private final PlaceVisitStorage placeStorage;
+  private final Gson gson;
+
+  private static final Pattern TRIP_NAME_PATTERN = Pattern.compile(".*/trips/([^/]+)/placeVisits");
+
+  private final GeoApiContext CONTEXT =
+      new GeoApiContext.Builder().apiKey("AIzaSyBbXCXC2uWv3baNmirLtqUYbFsFwCXqLV8").build();
+
+  public PlaceParentServlet(PlaceVisitStorage placeStorage, Gson gson) {
+    this.placeStorage = placeStorage;
+    this.gson = gson;
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    try {
+      Optional<PlaceVisitModel> place =
+          ServletUtil.extractFromRequestBody(request, gson, PlaceVisitModel.class);
+
+      String tripId = ServletUtil.parseUri(request, TRIP_NAME_PATTERN).group(1);
+      logger.atInfo().log("finished matching");
+
+      if (!place.isPresent()) {
+        response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        return;
+      }
+
+      PlaceDetailsRequest placeRequest =
+          new PlaceDetailsRequest(CONTEXT)
+              .placeId(place.get().placeId())
+              .fields(PlaceDetailsRequest.FieldMask.GEOMETRY_LOCATION,
+                  PlaceDetailsRequest.FieldMask.NAME);
+
+      PlaceDetails details;
+      details = placeRequest.await();
+
+      PlaceVisitModel newPlace = PlaceVisitModel.builder()
+                                     .setTripId(tripId)
+                                     .setPlaceId(place.get().placeId())
+                                     .setUserMark(PlaceVisitModel.UserMark.YES)
+                                     .setName(details.name)
+                                     .setLatitude(details.geometry.location.lat)
+                                     .setLongitude(details.geometry.location.lng)
+                                     .build();
+
+      placeStorage.updateUserMarkOrAddPlaceVisit(newPlace, PlaceVisitModel.UserMark.YES);
+
+      response.setContentType("application/json");
+      byte[] convert =
+          (gson.toJson(placeStorage.getPlaceVisit(tripId, place.get().placeId()).get()))
```

Comment:
You really shouldn't be calling `.placeId().get()` here, as this will throw if the place ID hasn't been set. Instead of using exceptions to control the flow in this case, you should handle it explicitly. Or, if it really doesn't make sense for a place to _not_ have a place ID, then it should be non-optional.

If this was made optional specifically for the purpose of client create requests not having a place ID (this is one of the benefits of having a separate, explicit request type), then we should probably have some external place precondition verifier that's called on any place returned by the storage. This makes the intent clear and prevents surprises where we get an invalid trip place (with no ID) and an inscrutable exception here.